examples/whl.py

```diff
@@ -0,0 +1,120 @@
+"""Install from a .whl file.
+"""
+
+import argparse
+import codecs
+import contextlib
+import sys
+import zipfile
+
+from installer._compat import pathlib
+from installer._compat.typing import TYPE_CHECKING
+from installer.layouts import DistInfo
+from installer.records import RecordItem, parse_record_file, write_record_file
+
+if TYPE_CHECKING:
+    from typing import Any, ContextManager, Dict, IO, Iterator
+
+
+class ZipFileInstaller(object):
+    """Install a local wheel.
+    """
+
+    def __init__(self, name, distinfo, zf):
+        # type: (str, DistInfo, zipfile.ZipFile) -> None
+        self._name = name
+        self._distinfo = distinfo
+        self._zf = zf
+
+    @classmethod
+    @contextlib.contextmanager
+    def create(cls, name, wheel_path):
+        # type: (str, pathlib.Path) -> Iterator[ZipFileInstaller]
+        project_name, project_version, _ = wheel_path.stem.split("-", 2)
+        with zipfile.ZipFile(str(wheel_path)) as zf:
+            entry_names = (name.lstrip("/").split("/", 1)[0] for name in zf.namelist())
+            distinfo = DistInfo.find(project_name, project_version, entry_names)
+            yield cls(name, distinfo, zf)
+
+    @contextlib.contextmanager
+    def _open_adjacent_tmp_for_write(self, path, **kwargs):
+        # type: (pathlib.Path, Any) -> Iterator[IO[Any]]
+        path.parent.mkdir(parents=True, exist_ok=True)
+        temp = path.with_name("{}.tmp.{}".format(path.name, self._name))
+        with temp.open(**kwargs) as f:
+            yield f
+        temp.replace(path)
+
+    def _open_target_for_write(self, path, binary=False):
+        # type: (pathlib.Path, bool) -> ContextManager[IO[Any]]
+        if binary:
+            kwargs = {"mode": "wb"}
+        else:
+            kwargs = {"mode": "w", "encoding": "utf-8"}
+        return self._open_adjacent_tmp_for_write(path, **kwargs)
+
+    def _open_csv_for_write(self, path):
+        # type: (pathlib.Path) -> ContextManager[IO[Any]]
+        if sys.version_info < (3,):
+            kwargs = {"mode": "wb"}
+        else:
+            kwargs = {"mode": "w", "newline": "", "encoding": "utf-8"}
+        return self._open_adjacent_tmp_for_write(path, **kwargs)
+
+    def _install_record_item(self, item, directory):
+        # type: (RecordItem, pathlib.Path) -> None
+        with self._zf.open(str(item.path)) as f:
+            data = f.read()
+        item.raise_for_validation(data)
+        target = directory.joinpath(item.path)
```

Comment:
instead of join path why not just use the ``/`` operator?

Reply:
The slash operator is a mistake and should have never existed IMO.

Reply:
I strongly disagree. IMO makes it a **lot** more readable, this just reads badly to me.

Reply:
And **I** strongly disagree. Aside the readable perspective (which is kind of subjective), the div operator also presents traps on operator precedence, and encourages inefficient coding style (`foo / "bar" / "baz"`).

This is like ly not a place to have this discussion though. Can I propose we acknowledge the difference and leave it as-is? Also this might be totally moot anyway if we use our own path implementation as you proposed.

Reply:
Our own path implementation would only be used on python2.7; on laters would still use pathlib. Duno for me if we use it like this we might as well use the ``os.path`` 🤷 and avoid all the problems of ``pathlib`` (like not being available on python 2)

Reply:
I still want something similar to `PosixPurePath` for `RECORD` even if we use `os.path` in most places though. Both `RECORD` and zipfile require forward slashes, and using string paths to interact with them is a recipe for bugs on Windows.

If we make our own `Path` I’d use it everywhere instead of on Python 2. We’ll be returning this type for `RECORD`, and it’d be woefully confusing if the two implementations differ too much.

Maybe we can use `zipp.Path` for this? It’s meant for zipfile, obviously, but the `RECORD` use case is *extremely* similar.